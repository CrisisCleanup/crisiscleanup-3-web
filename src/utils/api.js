const ROOT = process.env.VUE_APP_API_BASE_URL;

const getApiUrl = (path) => {
  const rootUrl = new URL(path, ROOT).toString();
  const getUrl = (endpoint) => {
    if (endpoint === undefined) {
      return rootUrl.slice(0, rootUrl.length - 1);
    }
    const fullUrl = new URL(endpoint, rootUrl).toString();
    return fullUrl;
  };
  return getUrl;
};

export const AgentApi = (endpoint) => getApiUrl('agents/')(endpoint);
export const PhoneApi = (endpoint) => getApiUrl('phone/')(endpoint);
export const IDPApi = (endpoint) => getApiUrl('idp/')(endpoint);
