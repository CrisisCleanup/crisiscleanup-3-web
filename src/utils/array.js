export function groupBy(array, key) {
  const result = {};
  array.forEach((item) => {
    if (!result[item[key]]) {
      result[item[key]] = [];
    }
    result[item[key]].push(item);
  });
  return result;
}
