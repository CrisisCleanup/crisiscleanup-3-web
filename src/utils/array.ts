export function groupBy(array, key) {
  const result = {};
  for (const item of array) {
    if (!result[item[key]]) {
      result[item[key]] = [];
    }

    result[item[key]].push(item);
  }

  return result;
}

export function childrenBy(dict, key) {
  const array = dict.null;
  for (const item of array) {
    item.children = dict[item[key]];
  }

  return array;
}
