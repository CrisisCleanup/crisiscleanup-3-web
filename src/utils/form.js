// @flow

import _ from 'lodash';
import type { LocaleFormFieldsT } from '@/components/forms/LocaleForm.vue';

export const groupBy = (key) => (array) =>
  array.reduce((objectsByKeyValue, obj) => {
    const value = obj[key];
    objectsByKeyValue[value] = (objectsByKeyValue[value] || []).concat(obj);
    return objectsByKeyValue;
  }, {});

export const buildForm = (key, dict, array) => {
  // TODO: refactor after writing unit tests
  // eslint-disable-next-line no-restricted-syntax
  for (const item of dict[key]) {
    if (item.label_t) {
      array.push(item);
    }
    if (item.field_key in dict) {
      buildForm(item.field_key, dict, array);
    }
  }
};

export const nest = (
  items,
  key = null,
  link = 'field_parent_key',
  excluded = [],
) =>
  items
    .filter((item) => item[link] === key && !excluded.includes(item.field_key))
    .map((item) => ({ ...item, children: nest(items, item.field_key) }));

export const nestUsers = (items, key = null) => {
  return items
    .filter((item) => {
      if (item.referring_user) {
        return item.referring_user === key;
      }
      return false;
    })
    .map((item) => {
      return {
        ...item,
        label: `${item.first_name} ${item.last_name}`,
        children: nestUsers(items, item.id),
      };
    });
};

export const emailRegex =
  /^[a-zA-Z0-9.!#$%&â€™*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;

/**
 * Iterates over locale inputs and generates locale keys for each.
 * @param inputs - locale inputs.
 * @param prefix - locale key prefix.
 * @param base - locale key base value.
 * @returns {Dictionary<LocaleFormFieldsT>}
 */
export const makeLocaleInputs = ({
  inputs,
  prefix = '',
  base = '',
}: {
  inputs: string[],
  prefix: string,
  base: string,
}): LocaleFormFieldsT =>
  _.transform(
    inputs,
    (result, value) => {
      const [name, suffix = null] = _.split(value, ':');
      let _key = `${prefix}.${base}${
        _.isNil(suffix) ? '' : `_${suffix || ''}`
      }`;
      const keyParts = _.split(_key, ':');
      _key = keyParts.join('_');
      result[name] = {
        value,
        key: _key,
      };
    },
    {},
  );
