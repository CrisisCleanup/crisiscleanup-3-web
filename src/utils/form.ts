import _ from 'lodash';

type LocaleFormFieldsT = Record<
  string,
  {
    locale: string | undefined;
    value: string | undefined;
    key: string | undefined;
  }
>;

export const groupBy =
  <T extends Record<string, any>, K extends keyof T>(key: K) =>
  (array: T[]): Record<T[K], T[]> =>
    array.reduce((objectsByKeyValue, object) => {
      const value = object[key];
      objectsByKeyValue[value] = (objectsByKeyValue[value] || []).concat(
        object,
      );
      return objectsByKeyValue;
    }, {} as Record<T[K], T[]>);

export const buildForm = <T extends Record<string, any>, K extends keyof T>(
  key: K,
  dict: Record<K, T[]>,
  array: T[],
) => {
  // TODO: refactor after writing unit tests

  for (const item of dict[key]) {
    if (item.label_t && Boolean(item.field_key)) {
      array.push(item);
    }

    if (item.field_key in dict && Boolean(item.field_key)) {
      buildForm(item.field_key, dict, array);
    }
  }
};

export const nest = <T extends Record<string, any>, K extends keyof T>(
  items: T[],
  key: K | null = null,
  link = 'field_parent_key',
  excluded: K[] = [],
): Record<K, T & { children: T[] | null }>[] =>
  items
    .filter((item) => Boolean(item.field_key))
    .filter((item) => item[link] === key && !excluded.includes(item.field_key))
    .map((item) => ({ ...item, children: nest(items, item.field_key) }));

export const nestUsers = <T extends Record<string, any>, K extends keyof T>(
  items: T[],
  key: K | null = null,
): Record<K, T & { label: string; children: T[] | null }>[] => {
  return items
    .filter((item) => {
      if (item.referring_user) {
        return item.referring_user === key;
      }

      return false;
    })
    .map((item) => {
      return {
        ...item,
        label: `${item.first_name} ${item.last_name}`,
        children: nestUsers(items, item.id),
      };
    });
};

export const emailRegex =
  /^[\w.!#$%&â€™*+/=?^`{|}~-]+@[a-zA-Z\d-]+(?:\.[a-zA-Z\d-]+)*$/;

/**
 * Iterates over locale inputs and generates locale keys for each.
 * @param inputs - locale inputs.
 * @param prefix - locale key prefix.
 * @param base - locale key base value.
 * @returns {Dictionary<LocaleFormFieldsT>}
 */
export const makeLocaleInputs = ({
  inputs,
  prefix = '',
  base = '',
}: {
  inputs: string[];
  prefix: string;
  base: string;
}): LocaleFormFieldsT =>
  _.transform(
    inputs,
    (result, value) => {
      const [name, suffix = null] = _.split(value, ':');
      let _key = `${prefix}.${base}${
        _.isNil(suffix) ? '' : `_${suffix || ''}`
      }`;
      const keyParts = _.split(_key, ':');
      _key = keyParts.join('_');
      result[name] = {
        value,
        key: _key,
      };
    },
    {} as Record<string, any>,
  );
