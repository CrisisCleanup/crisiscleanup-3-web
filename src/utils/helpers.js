/**
 * Convert rem to pixels.
 * @param rem
 * @returns {number}
 */
export const convertRemToPixels = (rem) => {
  return rem * parseFloat(getComputedStyle(document.documentElement).fontSize);
};

/**
 * Simple formatting for amounts.
 * @param num
 * @returns {number}
 */
export function nFormatter(num) {
  if (!num) return 0;
  if (num >= 1000000000) {
    return `${(num / 1000000000).toFixed(1).replace(/\.0$/, '')}G`;
  }
  if (num >= 1000000) {
    return `${(num / 1000000).toFixed(1).replace(/\.0$/, '')}M`;
  }
  if (num >= 1000) {
    return `${(num / 1000).toFixed(1).replace(/\.0$/, '')}K`;
  }
  return num;
}
