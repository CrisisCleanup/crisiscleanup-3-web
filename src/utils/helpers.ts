import { useI18n } from 'vue-i18n';
import { i18n } from '../main';

/**
 * Convert rem to pixels.
 * @param rem
 * @returns {number}
 */
export const convertRemToPixels = (rem: number): number => {
  return (
    rem * Number.parseFloat(getComputedStyle(document.documentElement).fontSize)
  );
};

/**
 * Simple formatting for amounts.
 * @returns {number | string}
 * @param number_
 */
export function nFormatter(number_: number): number | string {
  if (!number_) return 0;
  if (number_ >= 1_000_000_000) {
    return `${(number_ / 1_000_000_000).toFixed(1).replace(/\.0$/, '')}G`;
  }

  if (number_ >= 1_000_000) {
    return `${(number_ / 1_000_000).toFixed(1).replace(/\.0$/, '')}M`;
  }

  if (number_ >= 1000) {
    return `${(number_ / 1000).toFixed(1).replace(/\.0$/, '')}K`;
  }

  return number_;
}

export function formatCmsItem(text: string): string {
  return text.replaceAll(/{(.*?)}/g, (m, translation) => {
    return i18n.global.t(translation);
  });
}

export function numeral(
  value: number,
  type: string | undefined = null,
): string {
  let formatter = new Intl.NumberFormat('en-US', {
    maximumFractionDigits: 1,
    minimumFractionDigits: 0,
  });
  if (type === 'currency') {
    formatter = new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD',
      maximumFractionDigits: 0,
      minimumFractionDigits: 0,
    });
  }

  if (type === 'percentage') {
    formatter = new Intl.NumberFormat('en-US', {
      style: 'percent',
      minimumFractionDigits: 0,
      maximumFractionDigits: 0,
    });
  }

  return formatter.format(value || 0);
}
