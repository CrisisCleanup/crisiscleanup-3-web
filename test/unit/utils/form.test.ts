import { describe, expect, test } from 'vitest';
import { faker } from '@faker-js/faker';
import { generateMockIncident, generateMockUser } from '../../helpers';
import { groupBy, buildForm, nest, nestUsers, EMAIL_REGEX } from '@/utils/form';

describe('utils > urls', () => {
  const data = [
    { id: 1, name: 'John', age: 24, secretKey: null },
    { id: 2, name: 'Jane', age: 24, secretKey: null },
    { id: 3, name: 'John', age: 26, secretKey: null },
    { id: 4, name: 'John', age: 26, secretKey: null },
    { id: 5, name: 'Jack', age: 26, secretKey: null },
  ];

  test('groupBy', () => {
    expect(groupBy('secretKey')(data)).toMatchInlineSnapshot(`
      {
        "null": [
          {
            "age": 24,
            "id": 1,
            "name": "John",
            "secretKey": null,
          },
          {
            "age": 24,
            "id": 2,
            "name": "Jane",
            "secretKey": null,
          },
          {
            "age": 26,
            "id": 3,
            "name": "John",
            "secretKey": null,
          },
          {
            "age": 26,
            "id": 4,
            "name": "John",
            "secretKey": null,
          },
          {
            "age": 26,
            "id": 5,
            "name": "Jack",
            "secretKey": null,
          },
        ],
      }
    `);
    expect(groupBy('name')(data)).toMatchInlineSnapshot(`
      {
        "Jack": [
          {
            "age": 26,
            "id": 5,
            "name": "Jack",
            "secretKey": null,
          },
        ],
        "Jane": [
          {
            "age": 24,
            "id": 2,
            "name": "Jane",
            "secretKey": null,
          },
        ],
        "John": [
          {
            "age": 24,
            "id": 1,
            "name": "John",
            "secretKey": null,
          },
          {
            "age": 26,
            "id": 3,
            "name": "John",
            "secretKey": null,
          },
          {
            "age": 26,
            "id": 4,
            "name": "John",
            "secretKey": null,
          },
        ],
      }
    `);
  });

  test('buildForm', () => {
    const seed = faker.seed(345);
    console.log('Running with faker seed', seed);
    const mockIncident = generateMockIncident();
    const mockFormFields = mockIncident.form_fields;
    const grouped = groupBy('field_parent_key')(mockFormFields);
    const r = [] as Array<Record<string, any>>;
    // Grouped items might be grouped by null if field_parent_key is null
    // see test for form > groupBy
    buildForm('null', grouped, r);
    expect(r).toMatchSnapshot();
  });

  test('nest', () => {
    const data = [
      { id: 1, name: 'Abc1', field_parent_key: 'a1', field_key: 'aa1' },
      { id: 2, name: 'Abc2', field_parent_key: 'a1', field_key: 'aa2' },
      { id: 3, name: 'Abc3', field_parent_key: 'a1', field_key: 'aa3' },
      { id: 4, name: 'Abc4', field_parent_key: null, field_key: 'aa4' },
      { id: 5, name: 'Abc5', field_parent_key: null, field_key: 'aa5' },
      { id: 6, name: 'Abc6', field_parent_key: null, field_key: 'aa6' },
      { id: 7, name: 'Abc7', field_parent_key: 'a2', field_key: 'aa7' },
      { id: 8, name: 'Abc8', field_parent_key: 'a2', field_key: 'aa8' },
    ];
    const r1 = nest(data);
    const r2 = nest(data, 'a2', 'field_parent_key', ['field_key', 'name']);
    const r3 = nest(data, 'a1');
    expect(r1).toMatchSnapshot();
    expect(r2).toMatchSnapshot();
    expect(r3).toMatchSnapshot();
  });

  test('nestUsers', () => {
    const seed = faker.seed(456);
    console.log('Running with faker seed', seed);
    const _data = Array.from({ length: 4 }, generateMockUser);
    // set id and referring_user to test nesting
    const data = _data.map((d, i) => ({
      ...d,
      id: i + 1,
      referring_user: i === 0 ? null : i,
    }));
    const r = nestUsers(data, 1);
    expect(r).toMatchSnapshot();
  });

  // TODO: Improve email regex and make this test pass
  test.fails('EMAIL_REGEX', () => {
    // See: https://gist.github.com/cjaoude/fd9910626629b53c4d25
    // See: http://www.ex-parrot.com/pdw/Mail-RFC822-Address.html
    const emailRegex = EMAIL_REGEX;
    const validEmails = [
      'email@example.com',
      'firstname.lastname@example.com',
      'email@subdomain.example.com',
      'firstname+lastname@example.com',
      'email@123.123.123.123',
      'email@[123.123.123.123]',
      '“email”@example.com',
      '1234567890@example.com',
      'email@example-one.com',
      '_______@example.com',
      'email@example.name',
      'email@example.museum',
      'email@example.co.jp',
      'firstname-lastname@example.com',
    ];
    const invalidEmails = [
      'plainaddress',
      '#@%^%#$@#$@#.com',
      '@example.com',
      'Joe Smith <email@example.com>',
      'email.example.com',
      'email@example@example.com',
      '.email@example.com',
      'email.@example.com',
      'email..email@example.com',
      'あいうえお@example.com',
      'email@example.com (Joe Smith)',
      'email@example',
      'email@-example.com',
      'email@example.web',
      'email@111.222.333.44444',
      'email@example..com',
      'Abc..123@example.com',
    ];

    // array snapshot for valid emails (should be true)
    // TODO: improve regex to match all valid emails
    expect(validEmails.map((email) => [email, emailRegex.test(email)]))
      .toMatchInlineSnapshot(`
        [
          [
            "email@example.com",
            true,
          ],
          [
            "firstname.lastname@example.com",
            true,
          ],
          [
            "email@subdomain.example.com",
            true,
          ],
          [
            "firstname+lastname@example.com",
            true,
          ],
          [
            "email@123.123.123.123",
            true,
          ],
          [
            "email@[123.123.123.123]",
            true,
          ],
          [
            "“email”@example.com",
            true,
          ],
          [
            "1234567890@example.com",
            true,
          ],
          [
            "email@example-one.com",
            true,
          ],
          [
            "_______@example.com",
            true,
          ],
          [
            "email@example.name",
            true,
          ],
          [
            "email@example.museum",
            true,
          ],
          [
            "email@example.co.jp",
            true,
          ],
          [
            "firstname-lastname@example.com",
            true,
          ],
        ]
      `);
    // array snapshot for invalid emails (should be false)
    // TODO: improve regex to not match all invalid emails
    expect(invalidEmails.map((email) => [email, emailRegex.test(email)]))
      .toMatchInlineSnapshot(`
        [
          [
            "plainaddress",
            false,
          ],
          [
            "#@%^%#$@#$@#.com",
            false,
          ],
          [
            "@example.com",
            false,
          ],
          [
            "Joe Smith <email@example.com>",
            false,
          ],
          [
            "email.example.com",
            false,
          ],
          [
            "email@example@example.com",
            false,
          ],
          [
            ".email@example.com",
            false,
          ],
          [
            "email.@example.com",
            false,
          ],
          [
            "email..email@example.com",
            false,
          ],
          [
            "あいうえお@example.com",
            false,
          ],
          [
            "email@example.com (Joe Smith)",
            false,
          ],
          [
            "email@example",
            false,
          ],
          [
            "email@-example.com",
            false,
          ],
          [
            "email@example.web",
            false,
          ],
          [
            "email@111.222.333.44444",
            false,
          ],
          [
            "email@example..com",
            false,
          ],
          [
            "Abc..123@example.com",
            false,
          ],
        ]
      `);
  });
});
