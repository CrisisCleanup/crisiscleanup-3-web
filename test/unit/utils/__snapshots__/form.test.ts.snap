// Vitest Snapshot v1

exports[`utils > urls > buildForm 1`] = `
[
  {
    "data_group": "caseInfo",
    "data_sensitivity": "orgSensitive",
    "field_key": "property_info",
    "field_parent_key": null,
    "help_t": "Voluptate assumenda aperiam.",
    "html_type": "checkbox",
    "if_selected_then_work_type": null,
    "invalidated_at": null,
    "is_read_only": false,
    "is_read_only_default": true,
    "is_required": true,
    "is_required_default": true,
    "label_t": "Transexual",
    "list_order": 15735,
    "order_label": 11443,
    "phase": 4,
    "placeholder_t": "voluptatum",
    "read_only_break_glass": false,
    "validation": "string",
    "values": null,
    "values_default_t": null,
  },
  {
    "data_group": "personalInfo",
    "data_sensitivity": "verySensitive",
    "field_key": "children_in_home",
    "field_parent_key": null,
    "help_t": null,
    "html_type": "textarea",
    "if_selected_then_work_type": null,
    "invalidated_at": null,
    "is_read_only": true,
    "is_read_only_default": true,
    "is_required": false,
    "is_required_default": true,
    "label_t": "Toyota",
    "list_order": 50933,
    "order_label": 2403,
    "phase": 4,
    "placeholder_t": null,
    "read_only_break_glass": false,
    "validation": "string",
    "values": null,
    "values_default_t": null,
  },
  {
    "data_group": "workInfo",
    "data_sensitivity": "public",
    "field_key": "first_responder",
    "field_parent_key": null,
    "help_t": null,
    "html_type": "text",
    "if_selected_then_work_type": null,
    "invalidated_at": null,
    "is_read_only": false,
    "is_read_only_default": true,
    "is_required": false,
    "is_required_default": true,
    "label_t": "solidly",
    "list_order": 63126,
    "order_label": 58735,
    "phase": 4,
    "placeholder_t": null,
    "read_only_break_glass": true,
    "validation": "string",
    "values": null,
    "values_default_t": null,
  },
  {
    "data_group": "personalInfo",
    "data_sensitivity": "sensitive",
    "field_key": "veteran",
    "field_parent_key": null,
    "help_t": null,
    "html_type": "checkbox",
    "if_selected_then_work_type": null,
    "invalidated_at": null,
    "is_read_only": false,
    "is_read_only_default": false,
    "is_required": true,
    "is_required_default": true,
    "label_t": "SMTP",
    "list_order": 10239,
    "order_label": 82197,
    "phase": 4,
    "placeholder_t": null,
    "read_only_break_glass": false,
    "validation": "string",
    "values": null,
    "values_default_t": null,
  },
  {
    "data_group": "personalInfo",
    "data_sensitivity": "public",
    "field_key": "primary_language",
    "field_parent_key": null,
    "help_t": null,
    "html_type": "textarea",
    "if_selected_then_work_type": null,
    "invalidated_at": null,
    "is_read_only": false,
    "is_read_only_default": true,
    "is_required": false,
    "is_required_default": false,
    "label_t": "whoa",
    "list_order": 71488,
    "order_label": 95193,
    "phase": 4,
    "placeholder_t": null,
    "read_only_break_glass": true,
    "validation": "string",
    "values": null,
    "values_default_t": null,
  },
  {
    "data_group": "personalInfo",
    "data_sensitivity": "orgSensitive",
    "field_key": "rebuild_details",
    "field_parent_key": null,
    "help_t": "Debitis dignissimos sint aliquam at fugit distinctio explicabo quis.",
    "html_type": "textarea",
    "if_selected_then_work_type": null,
    "invalidated_at": null,
    "is_read_only": true,
    "is_read_only_default": true,
    "is_required": true,
    "is_required_default": false,
    "label_t": "blue",
    "list_order": 11569,
    "order_label": 45334,
    "phase": 4,
    "placeholder_t": "molestias",
    "read_only_break_glass": false,
    "validation": "string",
    "values": null,
    "values_default_t": null,
  },
  {
    "data_group": "personalInfo",
    "data_sensitivity": "public",
    "field_key": "debris_status",
    "field_parent_key": null,
    "help_t": null,
    "html_type": "text",
    "if_selected_then_work_type": null,
    "invalidated_at": null,
    "is_read_only": false,
    "is_read_only_default": true,
    "is_required": false,
    "is_required_default": true,
    "label_t": "doloremque",
    "list_order": 27620,
    "order_label": 89249,
    "phase": 4,
    "placeholder_t": "modi",
    "read_only_break_glass": false,
    "validation": "string",
    "values": null,
    "values_default_t": null,
  },
  {
    "data_group": "caseInfo",
    "data_sensitivity": "public",
    "field_key": "prepared_by",
    "field_parent_key": null,
    "help_t": null,
    "html_type": "select",
    "if_selected_then_work_type": null,
    "invalidated_at": null,
    "is_read_only": true,
    "is_read_only_default": false,
    "is_required": true,
    "is_required_default": true,
    "label_t": "definition",
    "list_order": 8733,
    "order_label": 64093,
    "phase": 4,
    "placeholder_t": null,
    "read_only_break_glass": true,
    "validation": "string",
    "values": null,
    "values_default_t": null,
  },
  {
    "data_group": "personalInfo",
    "data_sensitivity": "sensitive",
    "field_key": "do_not_work_before",
    "field_parent_key": null,
    "help_t": "Sint iste tempore officia praesentium fuga.",
    "html_type": "select",
    "if_selected_then_work_type": null,
    "invalidated_at": null,
    "is_read_only": false,
    "is_read_only_default": true,
    "is_required": true,
    "is_required_default": true,
    "label_t": "brandish",
    "list_order": 36850,
    "order_label": 28211,
    "phase": 4,
    "placeholder_t": null,
    "read_only_break_glass": false,
    "validation": "string",
    "values": null,
    "values_default_t": null,
  },
  {
    "data_group": "workInfo",
    "data_sensitivity": "public",
    "field_key": "claim_status_report_info_end",
    "field_parent_key": null,
    "help_t": null,
    "html_type": "checkbox",
    "if_selected_then_work_type": null,
    "invalidated_at": null,
    "is_read_only": false,
    "is_read_only_default": false,
    "is_required": true,
    "is_required_default": true,
    "label_t": "Screen",
    "list_order": 84396,
    "order_label": 12177,
    "phase": 4,
    "placeholder_t": null,
    "read_only_break_glass": false,
    "validation": "string",
    "values": null,
    "values_default_t": null,
  },
]
`;

exports[`utils > urls > nest 1`] = `
[
  {
    "children": [],
    "field_key": "aa4",
    "field_parent_key": null,
    "id": 4,
    "name": "Abc4",
  },
  {
    "children": [],
    "field_key": "aa5",
    "field_parent_key": null,
    "id": 5,
    "name": "Abc5",
  },
  {
    "children": [],
    "field_key": "aa6",
    "field_parent_key": null,
    "id": 6,
    "name": "Abc6",
  },
]
`;

exports[`utils > urls > nest 2`] = `
[
  {
    "children": [],
    "field_key": "aa7",
    "field_parent_key": "a2",
    "id": 7,
    "name": "Abc7",
  },
  {
    "children": [],
    "field_key": "aa8",
    "field_parent_key": "a2",
    "id": 8,
    "name": "Abc8",
  },
]
`;

exports[`utils > urls > nest 3`] = `
[
  {
    "children": [],
    "field_key": "aa1",
    "field_parent_key": "a1",
    "id": 1,
    "name": "Abc1",
  },
  {
    "children": [],
    "field_key": "aa2",
    "field_parent_key": "a1",
    "id": 2,
    "name": "Abc2",
  },
  {
    "children": [],
    "field_key": "aa3",
    "field_parent_key": "a1",
    "id": 3,
    "name": "Abc3",
  },
]
`;

exports[`utils > urls > nestUsers 1`] = `
[
  {
    "accepted_terms": null,
    "accepted_terms_timestamp": null,
    "active_roles": [
      9,
    ],
    "children": [
      {
        "accepted_terms": null,
        "accepted_terms_timestamp": null,
        "active_roles": [
          4,
        ],
        "children": [
          {
            "accepted_terms": null,
            "accepted_terms_timestamp": null,
            "active_roles": [
              9,
            ],
            "children": [],
            "email": "Maybell.Bergstrom@yahoo.com",
            "files": [],
            "first_name": "Gisselle",
            "id": 4,
            "label": "Gisselle Zieme",
            "last_name": "Zieme",
            "last_sign_in_at": null,
            "lineage": [
              5,
              7,
              9,
              1,
            ],
            "mobile": "358-900-1920 x919",
            "organization": {
              "affiliates": [
                3,
                8,
              ],
              "all_affiliates_and_groups": [
                8,
                6,
                1,
                7,
                5,
                2,
                9,
                10,
                3,
              ],
              "id": 68407,
              "is_active": true,
              "name": "Casper Group",
              "roles": [
                3,
                10,
                9,
                5,
                4,
                1,
                7,
              ],
              "type_t": "orgType.survivor_client_services",
            },
            "pending_roles": [],
            "primary_language": 1,
            "referring_user": 3,
            "roles": [
              9,
            ],
            "secondary_language": null,
            "sign_in_count": 0,
            "social": null,
          },
        ],
        "email": "Javier.Kuhic50@yahoo.com",
        "files": [],
        "first_name": "Lura",
        "id": 3,
        "label": "Lura Beahan",
        "last_name": "Beahan",
        "last_sign_in_at": null,
        "lineage": [
          7,
          8,
        ],
        "mobile": "(454) 797-1608",
        "organization": {
          "affiliates": [
            9,
            6,
            3,
            7,
            5,
            10,
            8,
            2,
          ],
          "all_affiliates_and_groups": [
            7,
            6,
            4,
            9,
            1,
          ],
          "id": 87428,
          "is_active": false,
          "name": "McDermott - Jast",
          "roles": [
            4,
            2,
            10,
            5,
            3,
            6,
            1,
            9,
          ],
          "type_t": "orgType.survivor_client_services",
        },
        "pending_roles": [],
        "primary_language": 2,
        "referring_user": 2,
        "roles": [
          4,
        ],
        "secondary_language": null,
        "sign_in_count": 0,
        "social": null,
      },
    ],
    "email": "Kaden_Corkery@hotmail.com",
    "files": [],
    "first_name": "Cicero",
    "id": 2,
    "label": "Cicero Wolf",
    "last_name": "Wolf",
    "last_sign_in_at": null,
    "lineage": [
      7,
      2,
      6,
      10,
      1,
      9,
      4,
      8,
      5,
      3,
    ],
    "mobile": "907.368.9609 x263",
    "organization": {
      "affiliates": [
        6,
        1,
      ],
      "all_affiliates_and_groups": [
        2,
        7,
        10,
        4,
        5,
        3,
        9,
        6,
        8,
        1,
      ],
      "id": 28827,
      "is_active": false,
      "name": "Murazik and Sons",
      "roles": [
        3,
        10,
        9,
        1,
        7,
        6,
        2,
        5,
        4,
        8,
      ],
      "type_t": "orgType.survivor_client_services",
    },
    "pending_roles": [],
    "primary_language": 3,
    "referring_user": 1,
    "roles": [
      9,
    ],
    "secondary_language": null,
    "sign_in_count": 0,
    "social": null,
  },
]
`;
